src/System.Management.Automation/engine/Interop/Windows/GetDosDevice.cs

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#nullable enable
+
+#if !UNIX
+using System;
+using System.Buffers;
+using System.ComponentModel;
+using System.Management.Automation;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class Windows
+    {
+        [LibraryImport(PinvokeDllNames.QueryDosDeviceDllName, EntryPoint = "QueryDosDeviceW", StringMarshalling = StringMarshalling.Utf16, SetLastError = true)]
+        internal static partial int QueryDosDevice(Span<char> lpDeviceName, Span<char> lpTargetPath, uint ucchMax);
+
+        internal static string GetDosDevice(char deviceName)
+        {
+            // By default buffer size is set to 300 which would generally be sufficient in most of the cases.
+            const int StartLength =
+#if DEBUG
+                // In debug, validate ArrayPool growth.
+                1;
+#else
+                300;
+#endif
+
+            Span<char> buffer = stackalloc char[StartLength + 1];
+            Span<char> fullDeviceName = stackalloc char[3] { deviceName, ':', '\0' };
+            char[]? rentedArray = null;
+
+            try
+            {
+                while (true)
+                {
+                    uint length = (uint)buffer.Length;
+                    int retValue = QueryDosDevice(fullDeviceName, buffer, length);
+                    if (retValue > 0)
+                    {
+                        if (buffer.StartsWith("\\??\\"))
+                        {
+                            // QueryDosDevice always return array of NULL-terinating strings with additional final NULL
+                            // so the buffer has always two NULL-s on end.
+                            //
+                            // "\\??\\UNC\\localhost\\c$\\tmp\0\0" -> "UNC\\localhost\\c$\\tmp\0\0"
+                            Span<char> res = buffer.Slice(4);
+                            if (res.StartsWith("UNC"))
+                            {
+                                // -> "C\\localhost\\c$\\tmp\0\0" -> "\\\\localhost\\c$\\tmp"
+                                res = res.Slice(2, retValue - 2);
```

Comment:
Per the docs https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-querydosdevicew, even for querying a specifying device, you can get multiple results:

>If lpDeviceName is non-NULL, the function retrieves information about the particular MS-DOS device specified by lpDeviceName. The first null-terminated string stored into the buffer is the current mapping for the device. The other null-terminated strings represent undeleted prior mappings for the device.

So it seems you need to find the first NULL rather than assume it only returns a single result.

Reply:
Good catch! 

Fixed.